### PR TITLE
Allow offline sessions in the middlewares

### LINKED ISSUE #58: Offline mode - Endless redirects
# Issue summary

When using v4 and setting accessMode 'offline', after a successful authentication the application will not redirect properly to the application home page. Instead it redirects to the 'fallbackRoute' and starts performing endless redirects to itself.

## Expected behavior

Prior to v4, accessMode 'offline' still loaded the application correctly after authentication. Should we be using online mode exclusively now and then requesting an offline token separately after a successful authentication?

### PR DESCRIPTION
**Pull request by paulomarg**
### WHY are these changes introduced?

Fixes #58

We were not handling the OAuth callback properly for offline tokens because we always expected online sessions to be used.

### WHAT is this pull request doing?

Making sure that we allow offline mode both in `shopifyAuth` and `verifyRequest`.

## Type of change

- [X] Patch: Bug (non-breaking change which fixes an issue)
- [ ] Minor: New feature (non-breaking change which adds functionality)
- [ ] Major: Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have added a changelog entry, prefixed by the type of change noted above
- [X] I have added/updated tests for this change

### CODE REVIEW

File: src/verify-request/verify-request.ts [hunk 1]
```diff
@@ -3,13 +3,21 @@ import compose from 'koa-compose';
 import {loginAgainIfDifferentShop} from './login-again-if-different-shop';
 import {verifyToken} from './verify-token';
 import {Options, Routes} from './types';
+import {DEFAULT_ACCESS_MODE} from '../auth';
 
 export default function verifyRequest(givenOptions: Options = {}) {
+  const {accessMode} = {
+    accessMode: DEFAULT_ACCESS_MODE,
+    ...givenOptions
+  };
```
Comment (mkevinosullivan) [inline src/verify-request/verify-request.ts:12]: I've not seen this before - if `...givenOptions` contains `accessMode`, it overwrites the `accessMode` that's set earlier in the object?

File: src/auth/index.ts [hunk 2]
```diff
@@ -96,7 +96,7 @@ export default function createShopifyAuth(options: OAuthStartOptions) {
       try {
         await Shopify.Auth.validateAuthCallback(ctx.req, ctx.res, ctx.query);
 
-        ctx.state.shopify = await Shopify.Utils.loadCurrentSession(ctx.req, ctx.res);
+        ctx.state.shopify = await Shopify.Utils.loadCurrentSession(ctx.req, ctx.res, config.accessMode === 'online');
```
Comment (mkevinosullivan) [inline src/auth/index.ts:99]: Is there any benefit to changing the `AccessMode` type to an `enum`?  Being able write this as
```suggestion
        ctx.state.shopify = await Shopify.Utils.loadCurrentSession(ctx.req, ctx.res, config.accessMode === AccessMode.Online);
```
seems more maintenance friendly - using `online` and `offline` strings as checks seems like an opportunity to create errors in the future.

One step further - is there perhaps an opportunity to create an `isOnline(accessMode: AccessMode)` utility method?

File: src/verify-request/verify-request.ts (hunk 1 above)
Comment (paulomarg) [reply src/verify-request/verify-request.ts:12]: Yes, exactly, I believe it amounts to the same as
```ts
Object.assign({}, <defaults>, givenOptions);
```

File: src/auth/index.ts (hunk 2 above)
Comment (paulomarg) [reply src/auth/index.ts:99]: The enum looks nicer, but I don't know how likely it is that new modes will be added in the future. All of these suggestions make sense, but should we consider them in a separate issue / PR?

Comment (paulomarg) [reply src/auth/index.ts:99]: I created https://github.com/Shopify/koa-shopify-auth/issues/60 since that change is slightly bigger than it seems at first.

**Review by mkevinosullivan — APPROVED** [review]
🚀

**Pull request merged**